xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-api/src/main/java/org/xwiki/notifications/page/events/PageNotificationEventListenerUpdater.java

```diff
@@ -0,0 +1,108 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.xwiki.notifications.page.events;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.xwiki.bridge.event.ApplicationReadyEvent;
+import org.xwiki.component.annotation.Component;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.notifications.internal.page.PageNotificationEventDescriptorManager;
+import org.xwiki.notifications.page.PageNotificationEventDescriptor;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+
+import com.xpn.xwiki.internal.event.XObjectAddedEvent;
+import com.xpn.xwiki.internal.event.XObjectDeletedEvent;
+import com.xpn.xwiki.internal.event.XObjectEvent;
+import com.xpn.xwiki.internal.event.XObjectUpdatedEvent;
+import com.xpn.xwiki.objects.BaseObjectReference;
+
+/**
+ * This listener triggers {@link PageNotificationEventDescriptorManager} when needed in order to refresh the
+ * {@link org.xwiki.notifications.page.PageNotificationEvent} registered by the {@link PageNotificationEventDescriptor}.
+ *
+ * @version $Id$
+ * @since 9.5RC1
+ */
+@Component
+@Singleton
+@Named(PageNotificationEventListenerUpdater.NAME)
+public class PageNotificationEventListenerUpdater extends AbstractEventListener
+{
+    /**
+     * The listener name.
+     */
+    public static final String NAME = "Page Notification Event Listener Updater";
+
+    @Inject
+    private PageNotificationEventDescriptorManager pageNotificationEventDescriptorManager;
+
+    /**
+     * Constructs a DefaultPageNotificationEventUpdater.
+     */
+    public PageNotificationEventListenerUpdater()
+    {
+        super(NAME,
+                new ApplicationReadyEvent(),
+                new XObjectAddedEvent(),
+                new XObjectDeletedEvent(),
+                new XObjectUpdatedEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        if (event instanceof ApplicationReadyEvent) {
+            pageNotificationEventDescriptorManager.updateDescriptors();
+        } else if (event instanceof XObjectAddedEvent
+                || event instanceof XObjectDeletedEvent
+                || event instanceof XObjectUpdatedEvent) {
+            // Extract the BaseObjectReference to be able to inspect the XClassReference.
+            BaseObjectReference objectReference = getBaseObjectReference((XObjectEvent) event);
+            DocumentReference objectClassReference = objectReference.getXClassReference();
+
+            // Only interested in PageNotificationEventDescriptorClass XObjects
```

Comment:
It's possible to directly match the class you want in the list of events. See https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/event/CommentEventGeneratorListener.java#L62.